File: day16_state/guru/cpp/context.cpp
#include <iostream>
#include <memory>
#include <typeinfo>

#include "context.hpp"
#include "state.hpp"


// weak_from_this is used instead of shared_from_this to break circular reference.
void Context::TransitionTo(std::shared_ptr<State> state) {
  // +2 on the name is just a manipulation on c-string for this specific case to remove prefix.
  std::cout << "Context: Transition to " << typeid(*state).name() + 2 << ".\n";
  this->state_ = state;
  this->state_->SetContext(weak_from_this());
}

void Context::Request1() {
  this->state_->Handle1();
}

void Context::Request2() {
  this->state_->Handle2();
}

std::shared_ptr<Context> Context::CreateSharedWithInitState(std::shared_ptr<State> state) {
  auto context = std::make_shared<Context>();
  context->TransitionTo(state);
  return context;
}

File: day16_state/guru/py/context.py
import state
# To avoid circular import problem, can't use from state import State.

class Context:
    """
    The Context defines the interface of interest to clients. It also maintains
    a reference to an instance of a State subclass, which represents the current
    state of the Context.
    """

    def __init__(self, state: state.State) -> None:
        self.transition_to(state)

    def transition_to(self, state: state.State) -> None:
        """
        The Context allows changing the State object at runtime.
        """

        print(f'Context: Transition to {type(state).__name__}')
        # A reference to the current state of the Context.
        self._state = state
        self._state.context = self

    """
    The Context delegates part of its behavior to the current State object.
    """
    def request_1(self) -> None:
        self._state.handle_1()

    def request_2(self) -> None:
        self._state.handle_2()
